### PR TITLE
Wrapper script for batch running NEON sites, and NEON fixes

### LINKED ISSUE #1354: Wrapper: Batch run single point sites, including NEON
Modify or create a shell script that can be used to run all of the NEON sites (including spinup and 'transient' cases)? 
- [x] Test this at all sites for 5 years (mostly to look for forcing data issues). 
- [x] Options for what type of simulations to do including:
  - ADspinup (e.g 200 years), 
  - postADspinup (e.g. 50 years), & 
  - 'transient' (2018-2020, with high frequency output), starting from 2018 restart files
  - Matrix (SASU) ins't critical now, this could come in later 
  - This can draw from Keith’s work for PLUMBER2  `/glade/work/oleson/release-clm5.0.34/tools/PTCLM/run_clmtowers_plumber2.csh`
  - It can also build on the .xml changes for spinup here `/glade/work/wwieder/ctsm/ctsm1.0.dev110/cime/scripts/clm50bgcNEON_ABBY/README_ww.case`
- [x] This would be  facilitated if NEON data were easier to query to know the length of simulations we can run.
- [x] Versatility of this script should let users: 
  - Identify particular sites & years 
  - Run all sites for all years

- **This greater complexity isn't critical now**,
  - Flexibility would let users select options:
  - NEON vs. 'other' sites (e.g. Ameriflux towers where input data are already created and on disk, somewhere) 
  - SP vs. BGC simulations (to determine the length of spinup and model configuration).
      - Initial configuration would be NEON - BGC options

### LINKED ISSUE #1407: mkmapdata.sh is failing for a regional grid, expecting to use MPI
### Brief summary of bug

@swensosc and I found this problem working on running mapping files for his regional grid.

### General bug information

**CTSM version you are using:** ctsm5.1.dev042
**Does this bug cause significantly incorrect results in the model's science?** No
**Configurations affected:** Running mkmapdata.sh for a regional grid

### Important details of your setup / configuration so we can reproduce the bug

`/glade/work/swensosc/ctsm_code/ctsm/tools/mkmapdata/mkmapdata.sh --gridfile /glade/scratch/swensosc/mapping/scripgrid/SCRIPgrid_ctsm_western_US_c210622.nc --res western_US -v`

### Important output or errors that show the problem

> /glade/work/swensosc/ctsm_code/ctsm/tools/mkmapdata/mkmapdata.sh: line 135: mpiexec_mpt: command not found

### LINKED ISSUE #1429: Modified NEON surface datasets have errors
### Brief summary of bug

It looks like a few sites that use the modified NEON surface dataset end up with errors in the soil sand, clay, and ORGANIC profiles.  

### General bug information
A quick look at some sites suggest this is an issue for HEAL, PUUM, TOOL and maybe others?  We don't even have met data for most of these sites, so it's likely not a big issue at this point, but something to be aware of as we start running more sites

@negin513 confirmed this is not an issue with how we're handling `zbedrock`, but results from issues in the raw data from NEON.

**Does this bug cause significantly incorrect results in the model's science?** YES

@ekluzek I don't want to delay bringing in #1375 and #1411, so let's bring in these PRs and @negin513 can address this later.

**Configurations affected:** NEON tower simulations

### LINKED ISSUE #1446: Modified neon data: dimensions lsmlat and lsmlon are missing where zbedrock is updated.
### Brief summary of bug

For some neon sites where we updated the `zberock` variable, the updated `zbedrock` does not have dimensions `lsmlat` and `lsmlon` which causes some issues. 

 This is where the `zbedrock` gets updated in the script:

https://github.com/ESCOMP/CTSM/blob/84e970fe5125ab97b93e6937afaed4539ef2ac33/tools/contrib/modify_singlept_site_neon.py#L516-L523

I should be able to resolve this issue. 

### General bug information

** CTSM version you are using: ctsm5.1.dev049 

**Configurations affected:** Neon simulations where `zbedrock` is updated, which are the following:
ORNL, CLBJ, HARV, BONA, OAES, TOOL, BARR, HEAL, PUUM, RMNP, SJER

Thanks to @ekluzek for bringing this to my attention.

### LINKED ISSUE #1454: run_sys_tests with --rerun-existing-failures option gives confusing result when testid doesn't exist
### Brief summary of bug

When you give the wrong testid for --rerun-existing-failures, it reports an error that's confusing and not clear what the problem is.

### General bug information

**CTSM version you are using:** ctsm5.1.dev051

**Does this bug cause significantly incorrect results in the model's science?** No
**Configurations affected:** Just running run_sys_tests with --rerun-existing-failures option

### Important output or errors that show the problem

The result it gives looks like this and it isn't obvious what the problem is...

> Traceback (most recent call last):
>   File "./run_sys_tests", line 20, in <module>
>     main(cime_path=cime_path)
>   File "/fs/cgd/data0/erik/ctsm_neon/python/ctsm/run_sys_tests.py", line 76, in main
>     extra_create_test_args=args.extra_create_test_args)
>   File "/fs/cgd/data0/erik/ctsm_neon/python/ctsm/run_sys_tests.py", line 148, in run_sys_tests
>     _record_git_status(testroot, retry_final, dry_run)
>   File "/fs/cgd/data0/erik/ctsm_neon/python/ctsm/run_sys_tests.py", line 453, in _record_git_status
>     with open(git_status_filepath, 'w') as git_status_file:
> FileNotFoundError: [Errno 2] No such file or directory: '/scratch/cluster/erik/tests_ctsm51d52_na/SRCROOT_GIT_STATUS'

### LINKED ISSUE #1456: Missing files for test tools
### Brief summary of bug

Two files are missing

modify_data_YELL
subset_data_YELL

in test/tools/nl_files

### General bug information

**CTSM version you are using:** ctsm5.1.dev051
**Does this bug cause significantly incorrect results in the model's science?** No

**Configurations affected:** Just tools testing

### Details of bug

The smia1 and smia2 tools tests will fail because of these missing files

Should go back to ctsm5.1.dev049

### Important output or errors that show the problem

It gives an error running the tests...
TSMscript_tools.sh: error running subset_data.py, error= 0

The test.log gives an error that reports on the usage of the script, since there was nothing for it to do.

### PR DESCRIPTION
**Pull request by negin513**
### Description of changes
This is a wrapper script for running CTSM simulation for one or more neon sites using CIME case objects.

Based on @ekluzek and @jedwards4b suggestions, I rewrote the wrapper script in #1426 to use CIME case objects.

This script accomplishes the following:
* Downloads and parses neon listing file for the available neon data :
https://neon-ncar.s3.data.neonscience.org/listing.csv

* Create a base generic case and clone other cases from this generic base case.
* Enable the user to run in AD, Post-AD, and transient mode.

Currently, there is no case dependency or case status check in the code. Which means that the user cannot run Post-AD +AD + transient all at the same time.  If we add case dependency, then the code that checks the status of a case (e.g. spinup) every 2 seconds and once it is finished successfully submit the next phase (e.g. transient.)

Externals are also updated to fix some NEON issues.

New tools subdirectory site_and_regional and current scripts for handling single-point moved underneath. Also README files and tools testing changed for this move. Also tools test added for the new wrapper script talked about above. Fix an issue in subset_data.py where it couldn't be run from outside of a git directory because it checks for the git hash.

Changes so that CTSM will write restarts as signaled at end of simulation by the CMEPS mediator.

### Specific notes

Contributors other than yourself, if any: @wwieder , @danicalombardozzi , @jedwards4b

CTSM Issues Fixed (include github issue #):  
  Fixes #1354
  Fixes #1454 
  Fixes #1456
  Fixes #1407
  Fixes #1446
  Fixes #1429
  Some work on #1441 to create new tools subdirectory and move some things around as well as updating README files

Are answers expected to change (and if so in what way)? NO

Any User Interface Changes (namelist or namelist defaults changes)? No

This script has the following interface:
```
|------------------------------------------------------------------|
|---------------------  Instructions  -----------------------------|
|------------------------------------------------------------------|
This is a wrapper script for running CTSM simulation for one or more
neon sites. 

This script is only for neon site and we will develop a more general
code later.

This script first creates and builds a generic base case. 
Next, it will clone the base_case for different neon sites and run
types to reduce the need to build ctsm everytime. 

This script will do the following:
    1) Create a generic base case for cloning. 
    2) Make the case for the specific neon site(s).
    3) Make changes to the case, for:
        a. AD spinup
	b. post-AD spinup
        c. transient
    	#---------------
    	d. SASU or Matrix spinup --> not working
    4) Build and submit the case.
 
-------------------------------------------------------------------
Instructions for running on Cheyenne/Casper:
 
load the following into your local environment
    module load python
    ncar_pylib
 
To remove NPL from your environment on Cheyenne/Casper:
    deactivate
 
-------------------------------------------------------------------
To see the available options:
    ./run_neon.py --help
-------------------------------------------------------------------

positional arguments:
  {ad,postad,transient,sasu}
                        Four different ways to run this script.
    ad                  AD spin-up options
    postad              Post-AD spin-up options
    transient           Transient spin-up options
    sasu                Sasu spin-up options --not in CTSM yet

optional arguments:
  -h, --help            show this help message and exit
  --neon_site {ABBY,BARR,BART,BLAN,BONA,CLBJ,CPER,DCFS,DEJU,DELA,DSNY,GRSM,GUAN,HARV,HEAL,JERC,JORN,KONA,KONZ,LAJA,LENO,MLBS,MOAB,NIWO,NOGP,OAES,ONAQ,ORNL,OSBS,PUUM,RMNP,SCBI,SERC,SJER,SOAP,SRER,STEI,STER,TALL,TEAK,TOOL,TREE,UKFS,UNDE,WOOD,WREF,YELL} [{ABBY,BARR,BART,BLAN,BONA,CLBJ,CPER,DCFS,DEJU,DELA,DSNY,GRSM,GUAN,HARV,HEAL,JERC,JORN,KONA,KONZ,LAJA,LENO,MLBS,MOAB,NIWO,NOGP,OAES,ONAQ,ORNL,OSBS,PUUM,RMNP,SCBI,SERC,SJER,SOAP,SRER,STEI,STER,TALL,TEAK,TOOL,TREE,UKFS,UNDE,WOOD,WREF,YELL} ...]
                        4-letter neon site code.
  --surf_dir SURF_DIR   Directory of single point surface dataset. [default:
                        /glade/scratch/negins/single_point/]
  --out_dir OUT_DIR     Directory to write updated single point surface dataset.
                        [default: /glade/scratch/negins/neon_sims/]
  -a, --all             Flag for running all neon sites, instead of selected few sites.
                        [default: False]
  -d, --debug           Debug mode will print more information.
-----------------
  ad 
                        run ad case
  --ad_length AD_LENGTH
                        How many years to run AD spin-up [default: 200]

-----------------
  postad 
                        run postad case
  --postad_length POSTAD_LENGTH
                        How many years to run in post-AD mode [default: 100]
  --finidat FINIDAT_POSTAD
                        finidat file location from spinup step to start from. [default:
                        None]
-----------------
  transient 
                        run transient case
  --start_year START_YEAR
                        Start year for running CTSM simulation. [default: 2018]
  --end_year END_YEAR   End year for running CTSM simulation. [default: 2020]
  --finidat FINIDAT_TRANSIENT
                        finidat file location from spinup step to start from. [default:
                        None]

```

### CODE REVIEW

**Comment by wwieder** [comment]
@negin513 should we close #1426 now?

File: tools/contrib/neon/neon_runner.py [hunk 1]
```diff
@@ -0,0 +1,793 @@
...
+ 
+#-- get the environment variable
+cesmroot = os.environ.get('CESM_ROOT')
+
+cesmroot = '/home/negins/ctsm_cime/'
+
+_LIBDIR = os.path.join(cesmroot,"cime","scripts","Tools")
+sys.path.append(_LIBDIR)
+_LIBDIR = os.path.join(cesmroot,"cime","scripts","lib")
+sys.path.append(_LIBDIR)
```
Comment (billsacks) [inline tools/contrib/neon/neon_runner.py:81]: I like the use of cime case objects here. Thanks for your work on this @negin513 !

Once this is integrated with the other ctsm python modules, you can make use of https://github.com/ESCOMP/CTSM/blob/master/python/ctsm/path_utils.py and/or https://github.com/ESCOMP/CTSM/blob/master/python/ctsm/add_cime_to_path.py to add the cime lib to your path in what I think is a robust way. (I'm open to suggestions if you can see a better way to do this.)

**Comment by wwieder** [comment]
After talking with @jedwards4b and @danicalombardozzi we decided that it's not worth the effort at this time to have the script automatically run AD, postAD, and transient simulations for single point cases. While this functionality is desirable in the long-term for CLM developers, it's not really critical for the NCAR-NEON application that we're focusing on now.

File: src/cpl/nuopc/lnd_comp_nuopc.F90 [hunk 2]
```diff
@@ -563,6 +566,12 @@ subroutine InitializeRealize(gcomp, importState, exportState, clock, rc)
     ! Set model clock in lnd_comp_shr
     model_clock = clock
 
+    call NUOPC_CompAttributeGet(gcomp, name="write_restart_at_endofrun", value=cvalue, isPresent=isPresent, isSet=isSet, rc=rc)
+    if (ChkErr(rc,__LINE__,u_FILE_u)) return
+    if (isPresent .and. isSet) then
+       if (trim(cvalue) .eq. '.true.') write_restart_at_endofrun = .true.
+    end if
+    print *,__FILE__,__LINE__,write_restart_at_endofrun
```
Comment (ekluzek) [inline src/cpl/nuopc/lnd_comp_nuopc.F90:574]: We'll remove this debug line.

File: src/cpl/nuopc/lnd_comp_nuopc.F90 [hunk 3]
```diff
@@ -1005,7 +1016,7 @@ subroutine ModelSetRunClock(gcomp, rc)
 
        call ESMF_GridCompGet(gcomp, name=name, rc=rc)
        if (ChkErr(rc,__LINE__,u_FILE_u)) return
-       call ESMF_LogWrite(subname//'setting alarms for ' // trim(name), ESMF_LOGMSG_INFO)
+       call ESMF_LogWrite(subname//'setting alarms for' // trim(name), ESMF_LOGMSG_INFO)
```
Comment (ekluzek) [inline src/cpl/nuopc/lnd_comp_nuopc.F90:1019]: We should keep the space after for...

**Review by ekluzek — APPROVED** [review]
There's a few small things that I'll go ahead and do that I make note of. I'll also do a first stab at making the gross directory structure we talked about in #1441. And I'm going to add a test/tools test for these new tools.

**Comment by ekluzek** [comment]
Standard testing has gone fine, with the exception of this test: ERS_Ly3_P72x2_Vnuopc.f10_f10_mg37.IHistClm50BgcCropG.cheyenne_intel.clm-cropMonthOutput
which fails at run time...

**Comment by jedwards4b** [comment]
@ekluzek  I think you are using the wrong externals here
cime should be cime6.0.5
cmeps should be cmeps0.13.22
cdeps is still to be tagged but will be cdeps0.12.19

**Comment by jedwards4b** [comment]
@ekluzek Erik please update cmeps to cmeps0.13.23.

**Comment by ekluzek** [comment]
OK I updated externals as advised but still had trouble with the above IG test (but now in CISM) until I backed off to cime6.0.4 where testing all worked as expected.

**Pull request merged**